### PR TITLE
Fuzzy Search

### PR DESCRIPTION
**Pull request by e82eric**
## Summary of the Pull Request
This pull request probably doesn't make sense to merge, but I found an open ticket for it and I had built a version for my personal build, so it felt like it was at least worth posting.  If you do decide to implement something like this I would love to help work on it.

https://github.com/microsoft/terminal/issues/5314

The fuzzy search implementation is from fzf-native. https://github.com/nvim-telescope/telescope-fzf-native.nvim
- Adapted to work with ICU/UText

Searching Terminal Repository
WindowsTerminal_1LQ4mBJENo

Searching .net memory dump
Fuzzy-cdb

## References and Relevant Issues
https://github.com/microsoft/terminal/issues/5314

## PR Checklist
- [ ] Closes #xxx
- [ ] Tests added/passed
- [ ] Documentation updated
   - If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/terminal) and link it here: #xxx
- [ ] Schema updated (if necessary)

### CODE REVIEW

**Comment by zadjii-msft** [comment]
but also, we're still wrapping up a couple internal commitments before we loop back to button up 1.20. I'm guessing just based on the scope of this delta that we may need to wait a bit before we can give this the review it very rightly deserves. Thanks for the patience here!

**Comment by DHowett** [comment]
_Holy s\*\*\*_!

**Comment by e82eric** [comment]
Thanks!  Totally understand, no rush on my end.

File: OpenConsole.sln [hunk 1]
```diff
@@ -2806,6 +2810,86 @@ Global
 		{328729E9-6723-416E-9C98-951F1473BBE1}.Release|ARM64.ActiveCfg = Release|ARM64
 		{328729E9-6723-416E-9C98-951F1473BBE1}.Release|x64.ActiveCfg = Release|x64
 		{328729E9-6723-416E-9C98-951F1473BBE1}.Release|x86.ActiveCfg = Release|Win32
+		{A52785FE-216E-4B64-9634-84172E083DB0}.AuditMode|Any CPU.ActiveCfg = Debug|x64
```
Comment (e82eric) [inline OpenConsole.sln:2813]: I am not sure how to keep visual studio from doing this.

File: src/cascadia/TerminalControl/TerminalControlLib.vcxproj.filters
Comment (e82eric) [inline src/cascadia/TerminalControl/TerminalControlLib.vcxproj.filters:1]: I think I can remove this

**Comment by zadjii-msft** [comment]
okay so what if we did this as a type of pane content, instead of as a part of the control itself? It could be like, the "Find All" pane or something (which is usually different in VsCode / sublime compared to the usual Find UI)

_this has been shower thoughts with mike_

**Comment by e82eric** [comment]
> okay so what if we did this as a type of pane content, instead of as a part of the control itself? It could be like, the "Find All" pane or something (which is usually different in VsCode / sublime compared to the usual Find UI)
> 
> _this has been shower thoughts with mike_

Trying to get my head around this,  by pane do you mean something like the docked areas in visual studio?

**Comment by zadjii-msft** [comment]
> by pane do you mean something like the docked areas in visual studio

Sorta, yea. In the chain of PRs that terminated in #16914, we added support for the Terminal to have panes that included things that weren't just TermControl's - they could be anything. 
image

So like, we could totally make a `FuzzyFindPaneContent` in the same manner. I'm sure there's some way we could like, stick a TermControl as a preview into that pane too. 

_(i cannot stress enough, this was a half baked shower thought)_

**Comment by e82eric** [comment]
> > by pane do you mean something like the docked areas in visual studio
> 
> Sorta, yea. In the chain of PRs that terminated in #16914, we added support for the Terminal to have panes that included things that weren't just TermControl's - they could be anything. image
> 
> So like, we could totally make a `FuzzyFindPaneContent` in the same manner. I'm sure there's some way we could like, stick a TermControl as a preview into that pane too.
> 
> _(i cannot stress enough, this was a half baked shower thought)_

got it,  it should work, that is similar to how I had it implemented at first, it was a xaml row below the swap panel control, and I was scrolling the swap panel to see the selected match (the scrolling wasn't necessary it was just nice to see the context rows around the selected match while in the list box).

the main downside was that when I was done with the fuzzy search I wanted the screen space back so I was hiding the row after the search, which mostly worked but cdb and a few other TUI applications would do weird things with the cursor with all the reflows (The reflows wouldn't be an issue if the pane doesn't auto hide).

having it in another pane would would make the code a lot simpler (mostly not having the extra renderer/swap panel for the previewer).

**Comment by e82eric** [comment]
@zadjii-msft  I am not sure if it helps,  this is what it looked like when I had the search control below the swap panel.  I think visually that would be similar to a pane (without the auto show/hide stuff)

fuzzy_pane

**Comment by zadjii-msft** [comment]
Honestly yea that looks almost exactly like what I was thinking. It's probably a lot of plumbing to get it up and into a separate `Pane`, but I'm thinking that might be a bit cleaner architecture-wise

**Comment by e82eric** [comment]
Makes sense,  I think the main thing is that the pane would need a reference to the Terminal instance (or some proxy/interface) to read the contents of the buffer, set the selected char and optionally perform scrolling to show the current item from the list box in the swap panel. 

I think selecting the char had a dependency on the Renderer also.

**Comment by zadjii-msft** [comment]
I don't think I'd be opposed to some winrt API's on `TermControl.idl` (and probably `ControlCore.idl`) to proxy through something like

```c#
struct SelectionRange {
  Windows.Foundation.Point Start;
  Windows.Foundation.Point End;
}

runtimeclass TermControl {
  // ...
  Vector<SelectionRange> GetPatternMatches(String pattern);

  void SelectRange(SelectionRange range); // basically just a projection of ControlCore::_selectSpan
}
```

**Comment by e82eric** [comment]
@zadjii-msft 

I can give it a try and see how it goes.
- I would use ScratchPad as an example?

Will this work for TermConrol?
```cpp
//HighlightedTextSegment is from the file below.  I am not sure if I would have access to it, if not I could create a similar set of clasess.
//I need the text spans to show the parts of the row that make up the fuzzy match
//https://github.com/microsoft/terminal/blob/32fbb16d43ddfcdd039a61c5af051a16b972faeb/src/cascadia/TerminalApp/HighlightedText.h#L11
struct FuzzySearchMetch {
  HighlightedTextSegment Text;
  //I think the start is the only part of match that is useful for selection since it is n number of spans across the row.  Selecting the first char in the match seems to get me where I want to go.  (This looked like how fzf and Telescope work)
  Windows.Foundation.Point Start;
}

runtimeclass TermControl {
  Vector<FuzzySearchMatch> GetPatternMatches(String pattern);
  void SelectPoint(SelectionPoint point);
}
```

For Vector<FuzzySearchMatch> GetPatternMatches(String pattern), These are the parts that I think I would need.
- Somewhere that has access to GetTextBuffer().  I think this would be ControlCore?
- This iterates over all of the rows in the textbuffer and scores them.
https://github.com/microsoft/terminal/blob/ae97bbd1819e5fa183125d30185a3d51dd343384/src/buffer/out/search.cpp#L37
- I needed this additional function in UTextAdapter to be able to support unicode, (before I was converting from wchar to char, I think this also keeps from copying the rows during the search)
https://github.com/microsoft/terminal/blob/ae97bbd1819e5fa183125d30185a3d51dd343384/src/buffer/out/UTextAdapter.cpp#L331
- This is the library that I used for the native implementation of FZF,  this may be a non starter for merging.  I think pretty much everyone the uses neovim ends using this, so it should be battle tested, but I don't know how it would fit with Microsoft's security policies etc.  https://github.com/nvim-telescope/telescope-fzf-native.nvim
- I adapted it to work with ICU/UText
https://github.com/microsoft/terminal/blob/ae97bbd1819e5fa183125d30185a3d51dd343384/src/buffer/fzf/fzf.c#L596

**Comment by e82eric** [comment]
@zadjii-msft I was able to get a really basic version of the fuzzy search working in the pane.

I forgot to ask,  was the idea that the fuzzy search would search across all TermControls?

This is what the diff looked like.
https://github.com/e82eric/terminal/commit/89a4e14661c747bafafaf40f98caf656634c2610

FuzzySearchPane

**Comment by zadjii-msft** [comment]
image

**Comment by zadjii-msft** [comment]
> I forgot to ask, was the idea that the fuzzy search would search across all TermControls

Let's save that for a v2? Cause this is already awesome

**Comment by e82eric** [comment]
@zadjii-msft  Any chance you guys are thinking about adding the ability to float a pane?  I was trying it out this weekend and it seemed pretty nice, for the fuzzy search it got out of the way after the search without having to reflow, for the TermControl it was nice to be able to do something temporary and have it get out the way (I actually think I found this more useful than the fuzzy search).

Fuzzy Search pane as float.
FloatFuzzy

TermControl pane as float.
FloatTerm

**Comment by zadjii-msft** [comment]
Ha you've figured out my plan! I was absolutely planning that (in a bit longer term).

We've got this QueryPalette element we're using currently for Terminal Chat:

image

and part of my eventual goal is to unify that floating "pane" with the rest of our Pane code, s.t. arbitrary panes could "float" like that too, or start as a floating pane, then get pinned as one in the tree, or.... who knows what yet. We still need to better sort out how some of those UI flows would work, but it's definitely something I've been thinking about.

**Comment by e82eric** [comment]
Awesome!  I surprised how nice it was having the termcontrol in a float windows.  

> I'm sure there's some way we could like, stick a TermControl as a preview into that pane too.

I was going to attempt adding a preview window using a TermControl just to see if it is possible or not, do I have it right that I would use the same instance of the ControlCore from main TermControl to create the preview TermControl?

**Comment by zadjii-msft** [comment]
> do I have it right that I would use the same instance of the `ControlCore` from main `TermControl` to create the preview `TermControl`

Yes, _ish_. Having multiple different TermControl's attached to the same ControlCore _at the same time_ is not something the codebase is super well suited for right now. I actually experimented with something like this just a couple months ago. I left my _very_ haphazard notes over in https://github.com/microsoft/terminal/issues/16495#issuecomment-1984745678. I'd warn that those notes are very rough and likely stale. But I think the basic idea - having a bunch of `TermControl`'s that can all connect to the same `ControlCore` - is probably pretty similar. 

(it's also probably way past overkill for a v0 of this PR, but something you could play with)

-----

I'll probably have more thoughts to share on this PR in like two weeks (after Build is done next week). Sorry for the delays!

**Comment by e82eric** [comment]
Thanks!  That gets me pointed in the right direction.

No rush on reviewing the PR.

**Comment by e82eric** [comment]
That worked.  (Adding the screenshot for reference,  agree that it is overkill for v0)
FuzzyPaneWithPreview

**Comment by zadjii-msft** [comment]
> I was able to get a really basic version of the fuzzy search working in the pane.
> 
> I forgot to ask, was the idea that the fuzzy search would search across all TermControls?
> 
> This is what the diff looked like. [e82eric@89a4e14](https://github.com/e82eric/terminal/commit/89a4e14661c747bafafaf40f98caf656634c2610)
> 

Looping back around - was there a reason this Pane version wasn't pushed as the tip of this PR? I think that's probably a lot closer to merging than all this IRenderData stuff. 

(I'm collecting up other notes too, but I wanted to make sure that question was answered first)

**Comment by e82eric** [comment]
> > I was able to get a really basic version of the fuzzy search working in the pane.
> > I forgot to ask, was the idea that the fuzzy search would search across all TermControls?
> > This is what the diff looked like. [e82eric@89a4e14](https://github.com/e82eric/terminal/commit/89a4e14661c747bafafaf40f98caf656634c2610)
> 
> Looping back around - was there a reason this Pane version wasn't pushed as the tip of this PR? I think that's probably a lot closer to merging than all this IRenderData stuff.
> 
> (I'm collecting up other notes too, but I wanted to make sure that question was answered first)

sorry,  I got distracted by the float plane and preview window stuff.  Will update the PR to use https://github.com/e82eric/terminal/commit/89a4e14661c747bafafaf40f98caf656634c2610 tonight.

**Comment by zadjii-msft** [comment]
> sorry, I got distracted by the float plane and preview window stuff. Will update the PR to use [e82eric@89a4e14](https://github.com/e82eric/terminal/commit/89a4e14661c747bafafaf40f98caf656634c2610) tonight.

No worries! Just wanted to make sure there wasn't something else I was missing. 

I'll write up the rest of our notes in the meantime ☺️

**Comment by e82eric** [comment]
> > sorry, I got distracted by the float plane and preview window stuff. Will update the PR to use [e82eric@89a4e14](https://github.com/e82eric/terminal/commit/89a4e14661c747bafafaf40f98caf656634c2610) tonight.
> 
> No worries! Just wanted to make sure there wasn't something else I was missing.
> 
> I'll write up the rest of our notes in the meantime ☺️

Should be updated now.

**Comment by zadjii-msft** [comment]
We talked about this as a team, and we're pretty excited about it. Especially the pane version, which seems a lot cleaner than before.
 
Our biggest concern right now is that we think that particular fzf implementation's license isn't quite right. The original fzf has a "Copyright (c) 2013-2024 Junegunn Choi" which is missing in the `telescope-fzf-native.vim` C port LICENSE. It just says "Copyright (c) 2021 Simon Hauser". Since it's a "derivative work" it's a license violation (it's a source port), because the 2nd paragraph in the MIT license states:
> The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

We can't fix that violation ourselves for the same reason, because the 2nd paragraph compels us to reproduce the license exactly, which we can, but it's an invalid one.
----> We need to contact `telescope-fzf-native.nvim` to fix their LICENSE file.
The right way to do it is to simply have 2 "Copyright ..." lines. For reference: https://softwareengineering.stackexchange.com/a/386584
 
We also wanna make sure that the fzf implementation handles non-ascii chars correctly, which we're not totally sure about. Polishing our support for unicode is one of our highest priorities right now - I'm not sure that we'd be able to take this out of canary-only until we can make sure that it'll better support unicode. (this is something we'll need to do diligence on our side)
 
I think @lhecker will have additional feedback on the particular `UText` integration into `fzf`. E.g. `ufzf_string_t` should just be a `UText*` ; `utextSlice_charAt` may return `U_SENTINEL `(= `0xffffffff`) which may be given to functions that can't deal with that = random behavior).
 
But big picture: this is super cool. Let's work together to get this across the finish line 😄

**Comment by e82eric** [comment]
Awesome,  I am excited to get to work on this.  Side note I have learned a ton from this and even if it doesn't get merged or out of the canary build it will have been time well spent for me.

Totally understand on the license. Is there anything I can do to help?

For the unicode stuff, any chance there is a list of chars/scenerios that need to be tested?  I had added a test project in the old PR that ported the telescope-native-fzf tests that I used to validate that stuff still worked after the Utext changes.  I also added some tests to see if it worked with some chars that I think were unicode,  (it looked like I mostly used \U0001F600: 😀 \u01C5: ǅ \u00CF: Ĩ).

If that type of test works I can add that test project to the new PR and try to add tests for all of the unicode scenerios.
https://github.com/e82eric/terminal/blob/ae97bbd1819e5fa183125d30185a3d51dd343384/src/buffer/fzf.Unit.Tests/fzf.Unit.Tests.cpp#L124

Will start looking at the ufzf_string_t/UText* stuff tonight.

**Comment by Paciupa** [comment]
Hi, excuse me. But what is the decision about this PR: It cannot be merged because of the license problem or it is been delayed? This looks too cool.

**Comment by Tq101514** [comment]
> ## Summary of the Pull Request
> 
> This pull request probably doesn't make sense to merge, but I found an open ticket for it and I had built a version for my personal build, so it felt like it was at least worth posting.  If you do decide to implement something like this I would love to help work on it.
> 
> 
> 
> https://github.com/microsoft/terminal/issues/5314
> 
> 
> 
> The fuzzy search implementation is from fzf-native. https://github.com/nvim-telescope/telescope-fzf-native.nvim
> 
> - Adapted to work with ICU/UText
> 
> 
> 
> Searching Terminal Repository
> 
> WindowsTerminal_1LQ4mBJENo
> 
> 
> 
> Searching .net memory dump
> 
> Fuzzy-cdb
> 
> 
> 
> ## References and Relevant Issues
> 
> https://github.com/microsoft/terminal/issues/5314
> 
> 
> 
> ## Detailed Description of the Pull Request / Additional comments
> 
> 
> 
> ## Validation Steps Performed
> 
> 
> 
> ## PR Checklist
> 
> - [ ] Closes #xxx
> 
> - [ ] Tests added/passed
> 
> - [ ] Documentation updated
> 
>    - If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/terminal) and link it here: #xxx
> 
> - [ ] Schema updated (if necessary)
> 
>

**Comment by lhecker** [comment]
> Hi, excuse me. But what is the decision about this PR: It cannot be merged because of the license problem or it is been delayed? This looks too cool.

My wish is for this PR to have a more robust fuzzy search implementation in respect to Unicode. The later changes already went into the right direction, but there's still a few concerning bits left in `fzf.c`. Honestly, it's primarily my failing to make time for this PR and to review it properly. 🥲
